File: sdk/kotlin/src/main/kotlin/com/loomup/client/JsonValue.kt
package com.loomup.client

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull

/** Flexible JSON value for dynamic table rows (v1 has no schema codegen). */
@Serializable(with = JsonValueSerializer::class)
sealed class JsonValue {
    data object Null : JsonValue()
    data class Bool(val value: Boolean) : JsonValue()
    data class Number(val value: Double) : JsonValue()
    data class String(val value: kotlin.String) : JsonValue()
    data class Array(val value: List<JsonValue>) : JsonValue()
    data class Object(val value: Map<kotlin.String, JsonValue>) : JsonValue()

    val stringValue: kotlin.String?
        get() = (this as? String)?.value

    val boolValue: Boolean?
        get() = (this as? Bool)?.value

    val numberValue: Double?
        get() = (this as? Number)?.value

    val intValue: Int?
        get() = (this as? Number)?.value?.toInt()

    val longValue: Long?
        get() = (this as? Number)?.value?.toLong()

    val objectValue: Map<kotlin.String, JsonValue>?
        get() = (this as? Object)?.value

    val arrayValue: List<JsonValue>?
        get() = (this as? Array)?.value

    operator fun get(key: kotlin.String): JsonValue? = objectValue?.get(key)

    companion object {
        fun from(element: JsonElement?): JsonValue {
            if (element == null || element is JsonNull) return Null
            return when (element) {
                is JsonPrimitive -> {
                    if (element.isString) {
                        String(element.content)
                    } else {
                        element.booleanOrNull?.let { Bool(it) }
                            ?: element.longOrNull?.let { Number(it.toDouble()) }
                            ?: element.doubleOrNull?.let { Number(it) }
                            ?: String(element.content)
                    }
                }
                is JsonArray -> Array(element.map { from(it) })
                is JsonObject -> Object(element.mapValues { from(it.value) })
                else -> Null
            }
        }

        fun fromAny(any: Any?): JsonValue {
            return when (any) {
                null -> Null
                is Boolean -> Bool(any)
                is kotlin.Number -> Number(any.toDouble())
                is kotlin.String -> String(any)
                is List<*> -> Array(any.map { fromAny(it) })
                is Map<*, *> -> {
                    val map = linkedMapOf<kotlin.String, JsonValue>()
                    for ((k, v) in any) {
                        if (k is kotlin.String) map[k] = fromAny(v)
                    }
                    Object(map)
                }
                else -> String(any.toString())
            }
        }
    }

    fun toJsonElement(): JsonElement = when (this) {
        is Null -> JsonNull
        is Bool -> JsonPrimitive(value)
        is Number -> {
            if (value == value.toLong().toDouble() &&
                value >= Long.MIN_VALUE.toDouble() &&
                value <= Long.MAX_VALUE.toDouble()
            ) {
                JsonPrimitive(value.toLong())
            } else {
                JsonPrimitive(value)
            }
        }
        is String -> JsonPrimitive(value)
        is Array -> buildJsonArray { value.forEach { add(it.toJsonElement()) } }
        is Object -> buildJsonObject {
            value.forEach { (k, v) -> put(k, v.toJsonElement()) }
        }
    }
}

object JsonValueSerializer : KSerializer<JsonValue> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("JsonValue")

    override fun serialize(encoder: Encoder, value: JsonValue) {
        val jsonEncoder = encoder as? JsonEncoder
            ?: error("JsonValue can only be serialized to JSON")
        jsonEncoder.encodeJsonElement(value.toJsonElement())
    }

    override fun deserialize(decoder: Decoder): JsonValue {
        val jsonDecoder = decoder as? JsonDecoder
            ?: error("JsonValue can only be deserialized from JSON")
        return JsonValue.from(jsonDecoder.decodeJsonElement())
    }
}

/** Scalar filter value for `select(where=...)`. */
sealed class WhereValue {
    data class String(val value: kotlin.String) : WhereValue()
    data class Int(val value: kotlin.Int) : WhereValue()
    data class Long(val value: kotlin.Long) : WhereValue()
    data class Double(val value: kotlin.Double) : WhereValue()
    data class Bool(val value: Boolean) : WhereValue()
    data class Many(val values: kotlin.collections.List<WhereValue>) : WhereValue()

    /** Query-string form. Booleans become `"1"` / `"0"` to match SQLite storage. */
    val queryString: kotlin.String
        get() = when (this) {
            is String -> value
            is Int -> value.toString()
            is Long -> value.toString()
            is Double -> value.toString()
            is Bool -> if (value) "1" else "0"
            is Many -> values.joinToString(",") { it.queryString }
        }

    companion object {
        fun of(value: kotlin.String): WhereValue = String(value)
        fun of(value: kotlin.Int): WhereValue = Int(value)
        fun of(value: kotlin.Double): WhereValue = Double(value)
        fun of(value: Boolean): WhereValue = Bool(value)
        fun of(value: kotlin.Long): WhereValue = Long(value)
        fun many(vararg values: WhereValue): WhereValue = Many(values.toList())
    }
}

File: swift/Sources/Loomup/JSONValue.swift
import Foundation

/// Flexible JSON value for dynamic table rows (v1 has no schema codegen).
public enum JSONValue: Sendable, Equatable, Hashable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    public var stringValue: String? {
        if case .string(let s) = self { return s }
        return nil
    }

    public var boolValue: Bool? {
        if case .bool(let b) = self { return b }
        return nil
    }

    public var numberValue: Double? {
        if case .number(let n) = self { return n }
        return nil
    }

    public var intValue: Int? {
        guard case .number(let n) = self else { return nil }
        return Int(n)
    }

    public var objectValue: [String: JSONValue]? {
        if case .object(let o) = self { return o }
        return nil
    }

    public var arrayValue: [JSONValue]? {
        if case .array(let a) = self { return a }
        return nil
    }

    public subscript(key: String) -> JSONValue? {
        objectValue?[key]
    }
}

extension JSONValue: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let b = try? container.decode(Bool.self) {
            self = .bool(b)
        } else if let i = try? container.decode(Int64.self) {
            self = .number(Double(i))
        } else if let d = try? container.decode(Double.self) {
            self = .number(d)
        } else if let s = try? container.decode(String.self) {
            self = .string(s)
        } else if let a = try? container.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? container.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Unsupported JSON value"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null:
            try container.encodeNil()
        case .bool(let b):
            try container.encode(b)
        case .number(let n):
            if n.rounded() == n, n >= Double(Int64.min), n <= Double(Int64.max) {
                try container.encode(Int64(n))
            } else {
                try container.encode(n)
            }
        case .string(let s):
            try container.encode(s)
        case .array(let a):
            try container.encode(a)
        case .object(let o):
            try container.encode(o)
        }
    }
}

extension JSONValue {
    /// Build from Foundation JSON object (`NSNull`, `NSNumber`, `String`, `Array`, `Dictionary`).
    public static func from(_ any: Any?) -> JSONValue {
        guard let any else { return .null }
        switch any {
        case is NSNull:
            return .null
        case let b as Bool:
            return .bool(b)
        case let n as NSNumber:
            // Distinguish Bool boxed as NSNumber on Apple platforms.
            let typeId = CFGetTypeID(n as CFTypeRef)
            if typeId == CFBooleanGetTypeID() {
                return .bool(n.boolValue)
            }
            return .number(n.doubleValue)
        case let s as String:
            return .string(s)
        case let a as [Any]:
            return .array(a.map { JSONValue.from($0) })
        case let d as [String: Any]:
            return .object(d.mapValues { JSONValue.from($0) })
        default:
            return .string(String(describing: any))
        }
    }

    public func toAny() -> Any {
        switch self {
        case .null: return NSNull()
        case .bool(let b): return b
        case .number(let n):
            if n.rounded() == n, n >= Double(Int64.min), n <= Double(Int64.max) {
                return Int64(n)
            }
            return n
        case .string(let s): return s
        case .array(let a): return a.map { $0.toAny() }
        case .object(let o): return o.mapValues { $0.toAny() }
        }
    }
}

/// Scalar filter value for `select(where:)`.
public enum WhereValue: Sendable, Equatable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case list([WhereValue])

    public init(_ value: String) { self = .string(value) }
    public init(_ value: Int) { self = .int(value) }
    public init(_ value: Double) { self = .double(value) }
    public init(_ value: Bool) { self = .bool(value) }
    public init(_ value: [WhereValue]) { self = .list(value) }

    /// Query-string form. Booleans become `"1"` / `"0"` to match SQLite storage.
    public var queryString: String {
        switch self {
        case .string(let s): return s
        case .int(let i): return String(i)
        case .double(let d): return String(d)
        case .bool(let b): return b ? "1" : "0"
        case .list(let values): return values.map(\.queryString).joined(separator: ",")
        }
    }
}

extension WhereValue: ExpressibleByStringLiteral {
    public init(stringLiteral value: String) { self = .string(value) }
}

extension WhereValue: ExpressibleByIntegerLiteral {
    public init(integerLiteral value: Int) { self = .int(value) }
}

extension WhereValue: ExpressibleByFloatLiteral {
    public init(floatLiteral value: Double) { self = .double(value) }
}

extension WhereValue: ExpressibleByBooleanLiteral {
    public init(booleanLiteral value: Bool) { self = .bool(value) }
}
